### PR TITLE
Unified stochastic kernel

### PR DESCRIPTION
**Pull request by habibutsu**
It is a draft of kernel.
For checking performance the following command can be used:
```bash
cargo test -p uzu --test kernel perf_batch -- --nocapture
```
on my laptop I have following results:
```
test sampling::sampling_perf_test::perf_batch1_128k_vocab ... [uzu::backends::metal::backend::Metal] batch=1 vocab=128000
  sequential: mean=0.312ms  min=0.298ms
  unified:    mean=0.137ms  min=0.134ms
  speedup:      2.23x
ok
test sampling::sampling_perf_test::perf_batch64_128k_vocab ... [uzu::backends::metal::backend::Metal] batch=64 vocab=128000
  sequential: mean=8.402ms  min=8.271ms
  unified:    mean=1.921ms  min=1.736ms
  speedup:      4.77x
ok
```

### CODE REVIEW

File: crates/uzu/src/session/parameter/sampling_policy.rs [hunk 1]
```diff
@@ -3,12 +3,21 @@ use crate::{config::LanguageModelConfig, session::parameter::ConfigResolvableVal
...
         min_p: Option<f32>,
     },
+    /// Unified single-pass path: all filtering and Gumbel-max in one kernel dispatch,
+    /// operating on logits loaded into private registers.
+    UnifiedStochastic {
+        temperature: Option<f32>,
+        top_k: Option<u32>,
+        top_p: Option<f32>,
+        min_p: Option<f32>,
+    },
```
Comment (uuuvn) [inline crates/uzu/src/session/parameter/sampling_policy.rs]: This should not be a separate sampling policy. Unified stochastic is an implementation detail, not a different sampling policy.

File: crates/uzu/src/backends/common/kernel/sampling.rs [hunk 2]
```diff
@@ -161,6 +166,41 @@ impl<B: Backend> SamplingKernel<B> {
...
+            if let Some(bitmask_buffer) = bitmask_buffer {
+                self.bitmask.encode(
+                    None::<&B::Buffer>,
+                    (bitmask_buffer, bitmask_offset),
+                    logits_buffer.deref_mut(),
+                    batch_size as u32,
+                    vocab_size as u32,
+                    command_buffer,
+                );
+            }
```
Comment (uuuvn) [inline crates/uzu/src/backends/common/kernel/sampling.rs]: This should also be fused

File: crates/uzu/src/backends/common/kernel/sampling.rs [hunk 3]
```diff
@@ -36,13 +36,15 @@ enum ArgmaxImplementation<B: Backend> {
...
+
 pub struct SamplingKernel<B: Backend> {
     bitmask: <B::Kernels as Kernels>::BitmaskKernel,
     temperature: <B::Kernels as Kernels>::TemperatureKernel,
     topk: <B::Kernels as Kernels>::TopKKernel,
     topp: <B::Kernels as Kernels>::TopPKernel,
     minp: <B::Kernels as Kernels>::MinPKernel,
     gumbel: <B::Kernels as Kernels>::GumbelKernel,
+    unified: <B::Kernels as Kernels>::UnifiedStochasticKernel,
     argmax_implementation: ArgmaxImplementation<B>,
```
Comment (uuuvn) [inline crates/uzu/src/backends/common/kernel/sampling.rs:41]: Are there any cases where separate kernels are faster? If no, unified stochastic should replace all of the old kernels, not just be a new option

File: crates/cli/src/main.rs [hunk 4]
```diff
@@ -26,6 +26,9 @@ enum Commands {
         message: Option<String>,
         #[arg(long, short)]
         no_thinking: bool,
+        /// Sampling method (default: stochastic with model's generation config)
+        #[arg(long)]
+        sampler: Option<SamplerArg>,
```
Comment (uuuvn) [inline crates/cli/src/main.rs]: Implementation details (unified/separate sampling kernels) shouldn't be exposed in the cli

File: crates/uzu/src/backends/cpu/kernel/sampling/unified_stochastic.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+use half::{bf16, f16};
+use num_traits::Float;
+
+use crate::{ArrayElement, language_model::gumbel::uniform_float};
+
+// CPU implementation of FlashInfer-style dual-pivot rejection sampling
+// with logit-space pivots. Mirrors the Metal kernel logic.
+#[kernel(UnifiedStochastic)]
+#[variants(T, f32, f16, bf16)]
+pub fn unified_stochastic<T: ArrayElement + Float>(
```
Comment (uuuvn) [inline crates/uzu/src/backends/cpu/kernel/sampling/unified_stochastic.rs]: Cpu backend is meant as a reference, it shouldn't be doing anything fancy like unified stochastic. I think there should be a single SamplingKernel trait that is implemented in the most straightforward textbook way possible on cpu and with either unified metal kernel directly or with multiple private metal kernels on metal.

File: crates/uzu/src/backends/metal/kernel/sampling/unified_stochastic.metal [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+    return fmax(float(out[0]) * (1.0f / 4294967296.0f), 1e-37f);
+}
+
+// ── Unified stochastic sampling: temperature + top_k/p/min_p + sampling in one dispatch ──
+//
+// NOTE: No Gumbel noise, no argmax.
+//   Gumbel-max (add Gumbel noise to logits → argmax) is mathematically equivalent to
+//   inverse-transform sampling from the softmax distribution (draw u ~ U(0,1), find
+//   token at CDF position u). This kernel uses the latter: one uniform draw per round,
+//   located via a cooperative prefix-sum walk — no per-token noise, no full-vocab argmax.
```
Comment (uuuvn) [inline crates/uzu/src/backends/metal/kernel/sampling/unified_stochastic.metal]: We use gumbel with shared seed between speculator and llm sampling for increased acceptance rate

**Review by uuuvn — APPROVED** [review]

**Review by uuuvn — CHANGES_REQUESTED** [review]
Accidentally selected the previous review as approve where it should've been request changes, don't see a way to undo it

File: crates/uzu/tests/context_mode_test.rs [hunk 7]
```diff
@@ -0,0 +1,151 @@
+mod common;
```
Comment (uuuvn) [inline crates/uzu/tests/context_mode_test.rs:1]: Why are you adding this file? crates/uzu/tests/integration/session/chat_session/context_mode_test.rs

File: crates/uzu/tests/sampling_test.rs [hunk 8]
```diff
@@ -0,0 +1,478 @@
+mod common;
```
Comment (uuuvn) [inline crates/uzu/tests/sampling_test.rs:1]: Why are you adding this file? crates/uzu/tests/unit/encodable_block/sampling_test.rs

File: crates/uzu/tests/kernel/sampling/mod.rs [hunk 9]
```diff
@@ -0,0 +1 @@
+mod sampling_perf_test;
```
Comment (uuuvn) [inline crates/uzu/tests/kernel/sampling/mod.rs:1]: Why are you adding a whole new directory for a single file?

File: crates/uzu/src/backends/cpu/kernel/sampling/stochastic.rs [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    #[optional(has_bitmask)] bitmask: Option<*const u32>,
+    batch_size: u32,
+    vocab_size: u32,
+    temperature: f32,
+    top_k: u32,
+    top_p: f32,
+    min_p: f32,
+    #[specialize] has_bitmask: bool,
+) {
+    let _ = min_p;
```
Comment (uuuvn) [inline crates/uzu/src/backends/cpu/kernel/sampling/stochastic.rs:24]: ?

File: crates/uzu/src/backends/cpu/kernel/sampling/stochastic.rs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        vocab_size
+    } else {
+        top_k as usize
+    };
+    let top_p = if top_p <= 0.0 {
+        1.0f32
+    } else {
+        top_p
+    };
+    let bitmask_stride = (vocab_size + 31) / 32;
```
Comment (uuuvn) [inline crates/uzu/src/backends/cpu/kernel/sampling/stochastic.rs:36]: div_ceil

File: crates/uzu/src/backends/common/kernel/sampling.rs [hunk 12]
```diff
@@ -36,14 +33,14 @@ enum ArgmaxImplementation<B: Backend> {
     },
 }
 
+// StochasticKernel retains N_CANDIDATES=64 candidates; top_k must not exceed this.
+const MAX_TOP_K: u32 = 64;
+
 pub struct SamplingKernel<B: Backend> {
     bitmask: <B::Kernels as Kernels>::BitmaskKernel,
```
Comment (uuuvn) [inline crates/uzu/src/backends/common/kernel/sampling.rs:40]: Let's add bitmask to argmax instead of having a separate kernel only for argmax

File: crates/uzu/src/backends/common/kernel/sampling.rs [hunk 13]
```diff
@@ -58,6 +55,10 @@ pub enum SamplingError<B: Backend> {
     BatchSizeExceeded(usize, usize),
     #[error("Vocab size {0} exceeds maximum {1}")]
     VocabSizeExceeded(usize, usize),
+    #[error("Stochastic: top_k={0} exceeds N_CANDIDATES={MAX_TOP_K}")]
+    TopKTooLarge(u32),
```
Comment (uuuvn) [inline crates/uzu/src/backends/common/kernel/sampling.rs:59]: We should probably still support top_k > 64 via some fallback.

File: crates/uzu/src/backends/common/kernel/sampling.rs [hunk 14]
```diff
@@ -159,85 +152,50 @@ impl<B: Backend> SamplingKernel<B> {
...
-            );
-        }
-
         if let SamplingMethod::Stochastic {
             temperature,
             top_k,
             top_p,
             min_p,
-            processing_order,
+            ..
```
Comment (uuuvn) [inline crates/uzu/src/backends/common/kernel/sampling.rs:160]: You're silently ignoring processing order

**Review by uuuvn — CHANGES_REQUESTED** [review]

**Comment by uuuvn** [comment]
Please carefully read your diff before re-requesting review, there is a lot of things that are either obviously wrong or very dirty

**Comment by uuuvn** [comment]
Stale, feel free to open a new one if you're interesting in finishing this
